File: frontend/src/types/Content.ts
/* -------------------------------------------------------------------------- */
/*                                Content Enums                               */
/* -------------------------------------------------------------------------- */

export enum ContentType {
  Blog = "blog",
  Project = "project",
}
  
export enum ProjectStatus {
  UnderDevelopment = "under_development",
  Finished = "finished"
}

/* -------------------------------------------------------------------------- */
/*                             Content Interfaces                             */
/* -------------------------------------------------------------------------- */

// If updating update to new content helper and the validation methods when they are added

/* ------------------ Existing/Updatable Content Interfaces ----------------- */

export interface Content extends NewContent {
  id: number;
  created_at: Date;
  updated_at: Date;
}

export interface Project extends NewProject {
  id: number;
  content_type: ContentType;
}

export interface Blog extends NewBlog {
  id: number;
  content_type: ContentType;
}

export interface FullContent {
  base_content: Content;
  extra_content: {project: Project} | {blog: Blog}
}

/* ------------------------- New Content Interfaces ------------------------- */

export interface NewContent {
  content_type: ContentType;
  slug: string;
  title: string;
  content_desc?: string;
  body: string;
}

export interface NewBlog {
  tags?: Array<string>;
}

export interface NewProject {
  current_status: ProjectStatus;
}

export interface NewFullContent {
  new_base_content: NewContent;
  new_extra_content: {project: NewProject} | {blog: NewBlog}
}

/* -------------------------------------------------------------------------- */

File: backend/src/db/models/content.rs


//TODO make a model that is changeable that dont have
// times or id and are all option feilds
#[derive(Queryable, AsChangeset, Identifiable, Serialize, Deserialize, Debug)]
#[diesel(table_name = content)]
#[diesel(treat_none_as_null = true)]
pub struct Content {
    id: i32,
    pub content_type: String,
    slug: String,
    title: String,
    content_desc: Option<String>,
    body: String,
    created_at: DateTime<Utc>,
    updated_at: DateTime<Utc>,
}


#[derive(Serialize, Deserialize, Debug)]
pub struct FullContent {
    pub base_content: Content,
    pub extra_content: ExtraContent,
}

#[derive(Insertable, Deserialize, Serialize, Debug)]
#[diesel(table_name = content)]
pub struct NewContent {
    content_type: String,
    slug: String,
    title: String,
    content_desc: Option<String>,
    body: String,
}

#[derive(Deserialize, Serialize, Debug)]
pub struct NewFullContent {
    pub new_base_content: NewContent,
    pub new_extra_content: NewExtraContent,
} 